# 评估数据

未清理的数据：脏数据与杂乱数据
未清理数据分为两种：

- 脏数据，也称为低质量数据。低质量数据存在内容问题。(损坏、缺失、重复的数据）
- 杂乱数据，也称为不整洁数据。不整洁数据存在结构问题。（每次观察构成一行、每类观察单元构成一个表格）

**数据质量维度**
数据质量维度有助于在评估和清理数据时为我们提供思维指导。四个主要的数据质量维度分别是：

- 完整性：我们是否记录了所有内容？是否缺少记录？是否丢失某个行、列或单元格？
- 有效性：我们已经做了记录，但却无效，即它们不符合定义的模式。模式是一组定义的数据规则。这些规则可以是真实世界约定成俗的事实（例如身高不可能是负数）和表格约定成俗的属性（例如表中的唯一键）。
- 准确性：不准确的数据是有效的，但仍然是错误的。这些数据符合定义的模式，但仍然不正确。例如：每个患者体重被多记录了5磅。虽然有失偏颇，这些数据仍然是有效的，但并不理想。
- 一致性：不一致的数据是有效和准确的，但是指代同一件事情的正确方式有多个。最好确保表和/或表内表示相同数据的列中的数据具有一致性，即采用标准格式。

**整洁的数据集**一个整洁的数据集易于操作、构建模型和可视化，并有一个特定的结构：

- 每个变量是一列
- 每个观察是一行
- 每个类型的观察单位就是一张表格。

以下是你经常会在 pandas 中使用的程序评估方法：

* .head (DataFrame 和 Series)
* .tail (DataFrame 和 Series)
* .sample (DataFrame 和 Series)
* .info (仅限于 DataFrame)
* .describe (DataFrame 和 Series)
* .value_counts (仅限于 Series)
* 各种索引和筛选数据的方法 (.loc and bracket notation with/without boolean indexing, also .iloc)


In [1]:
import pandas as pd
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

In [2]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [3]:
# 查看数据类型和缺失值
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [4]:
#查看整体的重复值
patients.duplicated().sum()

0

In [5]:
# 通过名字查看重复值 eg.
patients.surname.value_counts()

Doe               6
Jakobsen          3
Taylor            3
Batukayev         2
Ogochukwu         2
Hueber            2
Tucker            2
Cindrić           2
Gersten           2
Berg              2
Liễu              2
Nilsen            2
Souza             2
Schiavone         2
Woźniak           2
Dratchev          2
Lâm               2
Collins           2
Correia           2
Kowalczyk         2
Bùi               2
Johnson           2
Tạ                2
Parker            2
Lương             2
Silva             2
Lund              2
Aranda            2
Kadyrov           2
Grímsdóttir       2
                 ..
Afanasyeva        1
Mai               1
Mata              1
Galvez            1
Tománková         1
Compagnon         1
Bonami            1
Marchesi          1
Baum              1
Bakos             1
Iwata             1
Vũ                1
Herman            1
Brodahl           1
Koivunen          1
Halldórsdóttir    1
Hopkins           1
Kos               1
Wiśniewska        1


In [6]:
#通过地址查看重复值
patients.address.value_counts()

123 Main Street               6
2778 North Avenue             2
2476 Fulton Street            2
648 Old Dear Lane             2
4018 Poplar Avenue            1
3209 Crowfield Road           1
3942 Jerome Avenue            1
707 Gateway Avenue            1
2691 Kessla Way               1
1690 Fannie Street            1
2418 Smith Street             1
1895 Hart Ridge Road          1
1813 Lindale Avenue           1
1370 Flint Street             1
3234 Sardis Station           1
4435 Poe Road                 1
2126 Pearl Street             1
1821 Virginia Street          1
2386 Linda Street             1
3450 Southern Street          1
4040 Linda Street             1
1786 Gerald L. Bates Drive    1
1116 Dog Hill Lane            1
241 Freshour Circle           1
3130 Jessie Street            1
4145 Fairfax Drive            1
1846 Joseph Street            1
2043 Jadewood Drive           1
4243 Hidden Meadow Drive      1
513 Duck Creek Road           1
                             ..
2246 Phe

In [7]:
#查看重复地址具体数据
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [8]:
# 查看体重值得分布
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [9]:
#发现异常值并通过索引找到所在的位置
patients.iloc[210]

patient_id                                            211
assigned_sex                                       female
given_name                                        Camilla
surname                                          Zaitseva
address                               4689 Briarhill Lane
city                                              Wooster
state                                                  OH
zip_code                                            44691
country                                     United States
contact         330-202-2145CamillaZaitseva@superrito.com
birthdate                                      11/26/1938
weight                                               48.8
height                                                 63
bmi                                                  19.1
Name: 210, dtype: object

In [10]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [11]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [12]:
sum(treatments.auralin.isnull()), sum(treatments.novodra.isnull())

(0, 0)

In [13]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

#### 质量
##### `patients` 表格
- 邮编是浮点，而不是字符串
- 邮编有时是四位数
- Tim Neudorf 的身高是 27 英尺，而不是 72 英尺
- 有时使用州的全称，其他用简称
- Dsvid Gustafsson
- 缺少人口统计信息 (地址列至联系方式列) ***(无法清理)***
- 错误的数据类型 (指定性别、州、邮编和出生日期列)
- 多个手机号格式
- John Doe 的默认数据
- Jakobsen、Gersten、Taylor 多条记录
- Zaitseva 体重单位是 kgs，不是 lbs 

##### `treatments` 表格
- 缺少 HbA1c 变化
- Auralin 和 Novodra 初始剂量和最终剂量中的字母 'u' 
- 小写形式的姓和名
- 缺少记录 (280 而不是 350)
- 错误的数据类型 (auralin 和 novodra 列)
- HbA1c 变化不准确 (首个 4s 误写为 9s)
- auralin 和 novodra 列中空值用破折号 (-) 表示

##### `adverse_reactions` 表格
- 小写形式的姓和名

#### 清洁度
-  `patients` 表格中的联系方式一列应该拆分为手机号和邮箱两列
-  `treatments` 表格两列中的三个变量 (治疗、初始剂量和最终剂量)
-  不良反应应是 `treatments` 表格的一部分
-  `patients` 表格中的姓和名两列复制到 `treatments` 和 `adverse_reactions` 表格中

## 清理

In [14]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()
treatments_cut = pd.read_csv('treatments_cut.csv')

### 缺失数据

#### `treatments`： 缺少记录 (280 而不是 350)

##### 定义
*将treatments中的数据添加到treatments中。注释：缺少的 `treatments` 记录存储在文件中，命名为 `treatments_cut.csv`*

##### 代码

In [15]:
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [16]:
treatments_clean = pd.concat([treatments_clean,treatments_cut],ignore_index=True)

##### 测试

In [17]:
treatments_clean.shape

(350, 7)

#### `treatments`：缺少 HbA1c 变化和 HbA1c 变化不准确 (首个 4s 误写为 9s)
*注释： "HbA1c 变化不准确 (首个 4s 误写为 9s)" 的观察属于准确度问题，而不是完整性问题。

##### 定义
重新计算hba1c_change值

##### 代码

In [18]:
treatments_clean.hba1c_change = treatments_clean.hba1c_end - treatments_clean.hba1c_start

##### 测试

In [19]:
treatments_clean.hba1c_change.sort_values(ascending=False)

275   -0.20
70    -0.24
237   -0.26
144   -0.27
156   -0.27
200   -0.27
111   -0.28
320   -0.28
186   -0.28
227   -0.29
331   -0.29
326   -0.29
266   -0.29
93    -0.29
207   -0.30
141   -0.30
347   -0.30
201   -0.30
161   -0.30
44    -0.30
339   -0.30
192   -0.30
61    -0.31
259   -0.31
66    -0.31
150   -0.31
305   -0.31
95    -0.31
218   -0.31
324   -0.32
       ... 
226   -0.47
289   -0.47
17    -0.48
73    -0.48
225   -0.48
287   -0.48
49    -0.48
269   -0.49
138   -0.49
245   -0.49
82    -0.49
32    -0.49
224   -0.49
252   -0.50
308   -0.50
79    -0.50
86    -0.51
124   -0.51
10    -0.51
304   -0.51
136   -0.51
39    -0.51
195   -0.51
327   -0.51
158   -0.52
273   -0.52
187   -0.53
51    -0.54
178   -0.54
80    -0.54
Name: hba1c_change, Length: 350, dtype: float64

In [20]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0         -0.43  
1         -0.47  
2         -0.43  
3         -0.35  
4         -0.32

### 清洁度

####  `patients` 表格中的联系方式一列包括两个变量：手机号和邮箱

##### 定义
使用正则表达式和 pandas 的 `str.extract` 方法，从 *联系方式* 一列提取变量*手机号* 和 *邮箱* 。完成后删除 *联系方式* 一列。

##### 代码

In [21]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] 表明数据集中的邮箱开头和结尾都是字母
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 表明我们参考的是一列，而不是一行
patients_clean = patients_clean.drop('contact', axis=1)

In [22]:
patients_clean.phone_number.sample(50)

408    +1 (989) 390-0285
43          307 712 0508
301         646-982-6609
5           386-334-5237
462         213-749-6958
135         440-385-5011
373         479-248-5663
110    +1 (214) 637-0855
413         313-341-7799
274         740-552-2444
35          718-501-0503
496         209 762 2320
400         920-892-2965
143         206-902-0473
9           718 795 9124
428         412-319-0903
146         219 391 7729
11          931 207 0839
114         917-256-5329
193         203-933-3979
436         703-547-0551
155         618-261-9896
119         832-457-6625
432         979 203 0438
183         909-355-9418
48          312-719-7238
224         802-614-0812
479         909 458 2515
328    +1 (612) 342-6065
501         360 443 2060
276    +1 (208) 264-9063
244           1234567890
238         215-528-2193
170         718-628-9500
272    +1 (937) 518-7238
363         562-452-5685
337         208-830-2415
382         989-822-1351
365         540-986-1369
448         216-502-3773


##### 测试

In [23]:
# 确认联系方式一列可以运行
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [24]:
patients_clean.phone_number.sample(25)

338    +1 (985) 335-8662
321         504 403 4615
301         646-982-6609
358         203 235 1076
482         361-693-4960
400         920-892-2965
12          434-509-2614
322         254-546-2728
143         206-902-0473
154         608 277 0146
333         260 623 1176
112    +1 (757) 885-5669
463         573 437 7334
360         678 412 3728
122         231-607-3625
131         304-438-2648
294         508 921 6327
196         512-738-2609
183         909-355-9418
10          906-478-8949
56     +1 (706) 497-0891
257                  NaN
308         631-875-3023
79          302-397-6035
366         605-250-4942
Name: phone_number, dtype: object

In [25]:
# 确认没有邮箱以整数开头 (正则表达式不能匹配整数)
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

#### `treatments` 表格两列中的三个变量 (治疗、初始剂量和最终剂量)

##### 定义
在 *auralin* 和 *novodra* 两列中加入 *治疗* 和 *剂量* 列 (这时剂量仍然包含初始剂量和最终剂量)。然后使用破折号 ' - ' 把剂量列切分，得到 *初始剂量* 和 *最终剂量* 两列。删除中间使用的 *剂量* 一列。

##### 代码

In [26]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0         -0.43  
1         -0.47  
2         -0.43  
3         -0.35  
4         -0.32

In [27]:
# 将treatments表格两列中的三个变量转化成三列，用melt函数，先转化成treatment和dose两列。
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'],
        value_vars=['auralin','novodra'],
       var_name='treatment',value_name='dose')
treatments_clean = treatments_clean.drop(treatments_clean.query('dose=="-"').index)

In [28]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20         -0.43   auralin   
3       skye  gormanston         7.97       7.62         -0.35   auralin   
6     sophia      haugen         7.65       7.27         -0.38   auralin   
7      eddie      archer         7.89       7.55         -0.34   auralin   
9       asia     woźniak         7.76       7.37         -0.39   auralin   

        dose  
0  41u - 48u  
3  33u - 36u  
6  37u - 42u  
7  31u - 38u  
9  30u - 36u

In [29]:
# 将dose列用破折号分开成两列，利用str.split函数
treatments_clean_dose = treatments_clean['dose'].str.split(' - ', expand=True).\
rename(columns={0: "dose_start", 1: "dose_end"})

In [30]:
# 将拆开后的两列合并进原先的数据表格，并剔除dose列,重新建立索引
treatments_clean = pd.merge(treatments_clean,treatments_clean_dose,left_index=True,right_index=True)
treatments_clean = treatments_clean.drop('dose',axis=1)
treatments_clean = treatments_clean.reset_index(drop=True)

#### 测试

In [31]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20         -0.43   auralin   
1       skye  gormanston         7.97       7.62         -0.35   auralin   
2     sophia      haugen         7.65       7.27         -0.38   auralin   
3      eddie      archer         7.89       7.55         -0.34   auralin   
4       asia     woźniak         7.76       7.37         -0.39   auralin   

  dose_start dose_end  
0        41u      48u  
1        33u      36u  
2        37u      42u  
3        31u      38u  
4        30u      36u

#### 不良反应应是 `treatments` 表格的一部分

##### 定义
将 *不良反应* 一列合并到 `treatments` 表格中，按照 *名* 和 *姓*进行合并。

##### 代码

In [32]:
adverse_reactions_clean.shape

(34, 3)

In [33]:
treatments_clean = pd.merge(treatments_clean,adverse_reactions_clean,how='left',
         left_on=['given_name','surname'],right_on=['given_name','surname'])

#### 测试

In [34]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20         -0.43   auralin   
1       skye  gormanston         7.97       7.62         -0.35   auralin   
2     sophia      haugen         7.65       7.27         -0.38   auralin   
3      eddie      archer         7.89       7.55         -0.34   auralin   
4       asia     woźniak         7.76       7.37         -0.39   auralin   

  dose_start dose_end adverse_reaction  
0        41u      48u              NaN  
1        33u      36u              NaN  
2        37u      42u              NaN  
3        31u      38u              NaN  
4        30u      36u              NaN

####  `patients` 表格中的姓和名两列复制到 `treatments` 和 `adverse_reactions`表格中，以及小写形式的姓和名 

##### 定义
我们不再需要不良反应表格，所以忽略这部分内容。在 `patients` 表格中，提取病人编号和姓名，然后把这些名字改为小写形式，加入到 `treatments` 表格中。然后删除治疗表格中的姓和名两列 (这样小写形式不再是问题了)。

##### 代码

In [35]:
patients_clean.given_name = patients_clean.given_name.str.lower()
patients_clean.surname = patients_clean.surname.str.lower()

In [36]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        zoe  wellish       576 Brown Bear Drive   
1           2       female     pamela     hill  2370 University Hill Road   
2           3         male        jae   debord      1493 Poling Farm Road   
3           4         male       liêm     phan        2335 Webster Street   
4           5         male        tim  neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi       phone_number                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1       334-515-7487        TimNeudorf@cuvox.de

In [37]:
patients_id = patients_clean[['patient_id','given_name','surname']]

In [38]:
treatments_clean = pd.merge(patients_id,treatments_clean,how='right',
         left_on=['given_name','surname'],right_on=['given_name','surname'])

In [39]:
treatments_clean = treatments_clean.drop(['given_name','surname'],axis=1)

In [40]:
treatments_clean.head()

patient_id  hba1c_start  hba1c_end  hba1c_change treatment dose_start  \
0         1.0         7.71       7.30         -0.41   novodra        33u   
1         2.0         9.53       9.10         -0.43   novodra        27u   
2         4.0         7.58       7.10         -0.48   novodra        43u   
3         6.0         7.73       7.34         -0.39   auralin        50u   
4         7.0         7.65       7.26         -0.39   novodra        32u   

  dose_end adverse_reaction  
0      33u              NaN  
1      29u              NaN  
2      48u              NaN  
3      60u              NaN  
4      33u              NaN

#### 测试

In [41]:
# 病人编号应该是唯一复制的一列
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

15    patient_id
dtype: object

### 质量

#### 邮编是浮点，而不是字符串，有时是四位数

##### 定义
Convert the zip code column's data type from a float to a string 使用 `astype` 把邮件一列的数据类型从浮点转换为字符串，使用字符串分割删除 '.0'。在四位数邮编前面加上 0。 

##### 代码

In [42]:
patients_clean.dtypes

patient_id        int64
assigned_sex     object
given_name       object
surname          object
address          object
city             object
state            object
zip_code        float64
country          object
birthdate        object
weight          float64
height            int64
bmi             float64
phone_number     object
email            object
dtype: object

In [43]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5,side='left',fillchar='0')

#### Tim Neudorf 的身高是 27 英尺，而不是 72 英尺

##### 定义
在 `patients` 表格中身高为 27 英尺 (仅有一个) 的一行身高，替换为72英尺。

##### 代码

In [44]:
patients_clean.height = patients_clean.height.replace(27, 72)

#### 有时使用州的全称，其他用简称

##### 定义
使用函数，将 California (加利福尼亚)、New York (纽约)、Illinois (伊利诺伊)、Florida (佛罗里达) 和 Nebrask (内布拉斯加) 等州的全称转变为简称。

##### 代码

In [45]:
patients_clean.state.value_counts()

California    36
TX            32
New York      25
CA            24
MA            22
NY            22
PA            18
GA            15
OH            14
Illinois      14
MI            13
Florida       13
OK            13
LA            13
NJ            12
VA            11
WI            10
IL            10
MS            10
FL             9
TN             9
AL             9
MN             9
IN             9
KY             8
NC             8
WA             8
MO             7
ID             6
KS             6
NV             6
SC             5
IA             5
CT             5
ND             4
AR             4
RI             4
CO             4
ME             4
Nebraska       4
AZ             4
MD             3
SD             3
WV             3
OR             3
DE             3
DC             2
NE             2
VT             2
MT             2
WY             1
NM             1
NH             1
AK             1
Name: state, dtype: int64

In [46]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        zoe  wellish       576 Brown Bear Drive   
1           2       female     pamela     hill  2370 University Hill Road   
2           3         male        jae   debord      1493 Poling Farm Road   
3           4         male       liêm     phan        2335 Webster Street   
4           5         male        tim  neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi       phone_number                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      72  26.1       334-515-7487        TimNeudorf@cuvox.de

In [47]:
patients_clean['state'] = patients_clean['state'].replace('California','CA')
patients_clean['state'] = patients_clean['state'].replace('Illinois','IL')
patients_clean['state'] = patients_clean['state'].replace('New York','NY')
patients_clean['state'] = patients_clean['state'].replace('New York','NY')
patients_clean['state'] = patients_clean['state'].replace('Florida','FL')
patients_clean['state'] = patients_clean['state'].replace('Nebraska','NE')

##### 测试

In [48]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
FL    22
MA    22
PA    18
GA    15
OH    14
OK    13
MI    13
LA    13
NJ    12
VA    11
WI    10
MS    10
MN     9
AL     9
IN     9
TN     9
NC     8
WA     8
KY     8
MO     7
KS     6
NE     6
NV     6
ID     6
IA     5
CT     5
SC     5
AZ     4
CO     4
AR     4
ME     4
ND     4
RI     4
WV     3
MD     3
SD     3
OR     3
DE     3
VT     2
MT     2
DC     2
NH     1
AK     1
NM     1
WY     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
在 `patients` 表格中含有名 'Dsvid' 的一行里替换为'David'。

##### 代码

In [49]:
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### 测试

In [50]:
patients_clean[patients_clean.surname == 'Gustafsson']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

#### 错误的数据类型 (指定性别、州、邮编和出生日期列) 和错误的数据类型 (auralin 和 novodra 列)，以及 Auralin 和 Novodra 初始剂量和最终剂量中的字母 'u'

##### 定义
将指定性别和州转化为分类数据类型。把出生日期转化为 datetime 数据类型。在初始剂量和最终剂量中删除字母 'u'，并把这些列转化为整数数据类型。 

##### 代码

In [51]:
# 转为分类数据类型
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# 转为 datetime 类型
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# 添加 u 并转为整数类型
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

In [52]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        503 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [53]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 8 columns):
patient_id          349 non-null float64
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
dose_start          350 non-null int64
dose_end            350 non-null int64
adverse_reaction    35 non-null object
dtypes: float64(4), int64(2), object(2)
memory usage: 24.6+ KB


#### 多个手机号格式

##### 定义
删除所有的 " "、 "-"、"(", ")" 和 "+"，存储所有没有格式的数字。如果手机号只有 10 位数，在手机号前添加 1 (我们需要国家编号)。

##### 代码

In [54]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '', regex=True).str.pad(11, fillchar='1')

##### 测试

In [55]:
patients_clean.phone_number.head(20)

0     19517199170
1     12175693204
2     14023636804
3     17326368246
4     13345157487
5     13863345237
6     17755335933
7     14087783236
8     18162659578
9     17187959124
10    19064788949
11    19312070839
12    14345092614
13    12054178095
14    19187123469
15    12762251955
16    15597657836
17    16788298578
18    14067752696
19    15305328397
Name: phone_number, dtype: object

#### John Doe 的默认数据

##### 定义
从 `patients` 表格中删除不可恢复的 John Doe 记录。

##### 代码

In [56]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### 测试

In [57]:
# 应该没有 Doe 记录
patients_clean.surname.value_counts()

doe             6
taylor          3
jakobsen        3
hueber          2
bùi             2
kowalczyk       2
lâm             2
schiavone       2
gersten         2
ogochukwu       2
collins         2
batukayev       2
liễu            2
tucker          2
cabrera         2
correia         2
tạ              2
nilsen          2
parker          2
grímsdóttir     2
cindrić         2
berg            2
kadyrov         2
lund            2
souza           2
lương           2
silva           2
dratchev        2
aranda          2
johnson         2
               ..
zetticci        1
thạch           1
iwata           1
hori            1
mcgregor        1
barlow          1
bouw            1
mustafa         1
milne           1
bilodeau        1
filatov         1
nicholls        1
sokołowska      1
eidem           1
herczegh        1
zimmerman       1
teichelmann     1
saenger         1
woldendorp      1
vaara           1
endrizzi        1
martinsson      1
nucci           1
wiśniewski      1
michalčíko

In [58]:
# 应该没有 123 Main Street 记录
patients_clean.address.value_counts()

123 Main Street               6
2778 North Avenue             2
2476 Fulton Street            2
648 Old Dear Lane             2
4018 Poplar Avenue            1
3209 Crowfield Road           1
3942 Jerome Avenue            1
707 Gateway Avenue            1
2691 Kessla Way               1
1690 Fannie Street            1
2418 Smith Street             1
1895 Hart Ridge Road          1
1813 Lindale Avenue           1
1370 Flint Street             1
3234 Sardis Station           1
4435 Poe Road                 1
2126 Pearl Street             1
1821 Virginia Street          1
2386 Linda Street             1
3450 Southern Street          1
4040 Linda Street             1
1786 Gerald L. Bates Drive    1
1116 Dog Hill Lane            1
241 Freshour Circle           1
3130 Jessie Street            1
4145 Fairfax Drive            1
1846 Joseph Street            1
2043 Jadewood Drive           1
4243 Hidden Meadow Drive      1
513 Duck Creek Road           1
                             ..
2246 Phe

#### Jakobsen、Gersten、Taylor 多条记录

##### 定义
从 `patients` 表格中删除  Jake Jakobsen、Pat Gersten 和 Sandy Taylor。这些是昵称，不应该出现在 `treatments` 表格中 (删除错误名字可以保证  `patients` 和 `treatments` 表格的一致性问题)。这些是复制后第二次出现。这些是唯一出现的非空值复制地址。

##### 代码

In [59]:
# 波浪符号表示否： http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### 测试

In [60]:
patients_clean[patients_clean.surname == 'Jakobsen']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

In [61]:
patients_clean[patients_clean.surname == 'Gersten']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

In [62]:
patients_clean[patients_clean.surname == 'Taylor']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [63]:
patients_clean.shape

(495, 15)

#### Zaitseva 体重单位是 kgs，不是 lbs 

##### 定义
使用 [高级索引](https://stackoverflow.com/a/44913631) 把姓是 Zaitseva 的行隔离出来，并把体重单元格中的 kg 改为 lbs。

##### 代码

In [64]:
weight_kg = patients_clean.weight.sort_values()[0]
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

##### 测试

In [65]:
# 48.8 不应该是最低的
patients_clean.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
1      118.8
168    118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
12     224.2
252    224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
88     227.7
352    227.7
428    227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 495, dtype: float64